src/rtpghi~.c

```diff
@@ -1,158 +1,243 @@
-{
-    ltfat_int M = x->M_pd;
-    
-    //ltfat_complex_s *c = x->c;
-    post("length of c[]: %d", (M) * sizeof *(x->c));
-    post("s_n: %d", sp[0]->s_n);
-    
-    t_sample **dummy=x->out;
-    *dummy++=sp[1]->s_vec;
-    *dummy=sp[2]->s_vec;
-    
-    dsp_add(rtpghi_tilde_perform, 3,x, sp[0]->s_vec, sp[0]->s_n);
+
+void rtpghi_tilde_recreatestate(t_rtpghi_tilde *x, LTFAT_FIRWIN win, ltfat_int blocksize, int overlap) {
+	if(x->sta_pd && (win != x->window || blocksize != x->blocksize || overlap != x->overlap)) {
+		rtpghi_done(&(x->sta_pd));
+		x->sta_pd=NULL;
+		post("sta zeroed");
+	}
+	if (!x->sta_pd) {
+		ltfat_int w = 1;
+		x->blocksize = blocksize;
+		x->window = win;
+		x->overlap = overlap;
+		double tol = x->tol_pd;
+		int do_causal = x->do_causal_pd;
+		double gamma;
+		
+		gamma = phaseret_firwin2gamma(win, x->blocksize);
+		if (rtpghi_init(w, blocksize/(x->overlap), x->blocksize, gamma, tol, do_causal, &(x->sta_pd))) {
+			pd_error(x, "failed to init state");
+			x->sta_pd=NULL;
+			}
+		}
 }
 
-void *rtpghi_tilde_new(t_symbol *s, int argc, t_atom *argv)
-{
+void rtpghi_tilde_dsp(t_rtpghi_tilde *x, t_signal **sp) {
+	//post("start DSP");
+	ltfat_int M = (ltfat_int) sp[0]->s_n;
+	if ((x->blocksize) == 0) {
+		(x->blocksize) = M;
+	}
+
+	if (x->window_type_pd != NULL) {
+		x->window = ltfat_str2firwin(x->window_type_pd->s_name);
+	}
 
-  t_rtpghi_tilde *x = (t_rtpghi_tilde *)pd_new(rtpghi_tilde_class);
+	rtpghi_tilde_recreatestate(x, x->window, x->blocksize, x->overlap);
+	
+	if (x->c != NULL) {
+		freebytes(x->c, M * sizeof *(x->c));
+		x->c = NULL;
+	}
+	if (x->c == NULL) {
+		x->c = getbytes(M * sizeof *(x->c));
+	}
+	
+	if (x->sta_pd == NULL) {
+	   pd_error(x, "no output processing due to initalisation failure");
+	}
+	dsp_add(rtpghi_tilde_perform, 5,x, sp[0]->s_vec, sp[1]->s_vec, sp[2]->s_vec, sp[0]->s_n);
+}
 
-    x->W_pd=atom_getfloat(argv);
-    x->ol_pd=atom_getfloat(argv+1);
-    x->M_pd=atom_getfloat(argv+2);
-    x->tol_pd=atom_getfloat(argv+3);
-    x->do_causal_pd=atom_getfloat(argv+4);
-    x->window_type= *atom_getsymbol(argv+5);
+void rtpghi_tilde_causal(t_rtpghi_tilde *x, t_floatarg f) {
+	
+	if ((f == 0.0) || (f == 1.0)) {
+		int causal = f;
+		rtpghi_set_causal(x->sta_pd, causal);
+	}
+	else {
+		pd_error(x, "failed to set (a)causal, input has to be either 0 or 1, but got: %f", f);
+	}
+}
+void rtpghi_tilde_tol(t_rtpghi_tilde *x, t_floatarg f) {
+	
+	if ((f > 0.0) && (f < 1.0)) {
+		double tol = f;
+		rtpghi_set_tol(x->sta_pd, tol);
+	}
+	else {
+		pd_error(x, "failed to set tol, input float has to be  > 0 && <1, but got: %f", f);
+	}
+}
+
+void rtpghi_tilde_overlap(t_rtpghi_tilde *x, t_floatarg f) {
+	if ((f > 0.0) && (f <= x->blocksize)) {
+		rtpghi_tilde_recreatestate(x, x->window, x->blocksize, f);
+	}
+	else {
+		pd_error(x, "failed to set overlap, input float has to be  > 0 && <= blocksize, but got: %f", f);
+	}
+}
+
+void rtpghi_tilde_window(t_rtpghi_tilde *x, t_symbol s) {
```

Comment:
`s` really is a pointer `t_symbol*` (which kind of explains why it is crashing for me)